README
furl
----

A small utility designed to display the HTTP headers returned by web-servers in response to client requests.

Usage: furl [OPTIONS] [URL] [filename]
Where URL is in the format:

      [http://]www.foobar.com[:80][/example.html]


Options currently include the following:

-f, --file     specify an input file containing URLs
-b, --body     output the contents of HTTP request and the header
-V, --version  output version information and exit
-h, --help     display this help message and exit
-i, --ie       pretend to be Internet Explorer
-m, --mozilla  pretend to be Mozilla


Compiling
---------

This version of furl uses the Autoconf and Automake. On UNIX-like hosts try the following:

$ ./configure
$ make
$ su
# make install

On Windows ( >= Windows 95 ) there are configuration files in the Win32 sub-directory for compiling with:

Mingw32 - mingw32.bat
Borland C++ - borland.bat
MS Visual C++ - vcmake.bat

Change directories to the Win32 sub-directory and run the applicable batch file.

It is also possible to cross-compile using Mingw32 on GNU/Linux by altering the Win32/Makefile.mingw32 file.

Kidney Bingos
chris@bingosnet.co.uk


ChangeLog
2004-06-08: Added options to allow furl to "emulate" Internet Explorer or Mozilla.


furl.c
/*
 * furl 
 * Copyright (C) 1999 BingosNET Produktions Ltd
 * Written by Kidney Bingos aka Chris Williams
 * <chris@kidney-bingos.demon.co.uk> 
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

#ifndef WIN32
#include <stdio.h>
#endif
#include <stdlib.h>
#include <string.h>
#include "getopt.h"
#include "furlutil.h"

int main(int argc, char *argv[])
{

	static struct option options[] = 
	{
		{"body", no_argument,NULL,'b'},
		{"version", no_argument,NULL,'V'},
		{"file", required_argument,NULL,'f'},
		{"help",    no_argument,NULL,'h'},
		{"ie", no_argument,NULL,'i'},
		{"mozilla", no_argument,NULL, 'm'},
	};

	FILE *in = 0;
	char opt;
	char *optstring = "imbVhf:";
	char *progname;
	int file_count, fileind, length;
	int file_open = 0;
	char protocol[1024];
	char port[20];
	char path[1024];
	char hostname[1024];
	char fileurl[1024];
	char method[5] = "HEAD";
	char useragent[100] = "Furl/2.1";

	/* Let's see if it works */
	unsigned long ipAddr;
	SOCKET sock;

#ifdef WIN32
	Win32Init();
#endif

	progname = basename(argv[0]);	
	strlwr(progname);

	while ((opt=getopt_long(argc,argv,optstring,options,NULL)) != EOF ) {
		switch (opt) {
			
			case 'i':
				strcpy(useragent, "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1;)");
				break;
			case 'm':
				strcpy(useragent, "Mozilla/5.0 (Windows; U; Windows NT 5.1; en-US; rv:1.7b) Gecko/20040421");
				break;
			case 'b':
				strcpy(method, "GET");
				break;
			case 'V':
				PrintVersion(progname);
				return 0;

			case 'h':
				PrintUsage(progname);
				return 0;

			case 'f':
				if((in=fopen(optarg, "r")) == NULL) {
					fprintf(stderr,"Bad Magic. Can\'t open specified file %s\n", optarg);
#ifdef WIN32
					Win32Cleanup();
#endif
					exit(1);
				} else {
					file_open = 1;
				}
				break;

			default:
				fprintf(stderr,"Try '%s --help' for more information.\n",progname);
				return 1;
		}
	}
	
	file_count = argc - optind;

	fileind = optind++;

	if((file_count == 0  || file_count > 1) && !file_open) {

		PrintUsage(progname);
#ifdef WIN32
		Win32Cleanup();
#endif
		exit(1);

	} else {

		while(1) {
			if (file_open) {
				if(fgets(fileurl,1022,in) == NULL) {
					if (feof(in)) {
						break;
					} else {
						perror("Bad Magic ");
#ifdef WIN32
						Win32Cleanup();
#endif
						exit(1);
					}
				} else {
					length = strlen(fileurl);
					fileurl[length-1] = '\0';
					fprintf(stdout,"\nURL= %s\n\n", fileurl);
				}	
			} else {
				length = -1;
				strcpy(fileurl, argv[fileind]);
			}
			if (length > 1 || length == -1) {
				if (!parseurl(fileurl,protocol,hostname,port,path,method,useragent)) {

					if (GetAddress(&ipAddr, hostname)) {
						fprintf(stderr,"%s doesn't exist\n",hostname);
						fprintf(stderr,"Try '%s --help' for more information.\n",progname);
					} else {
						sock = CreateSocket(&ipAddr,port);
						if (sock == INVALID_SOCKET) {
							fprintf(stderr,"Bad Magic. Couldn\'t get a socket connection\n");
						} else {
							if (ProcessSocket(sock, path) == SOCKET_ERROR) {
								fprintf(stderr,"Bad Magic. Encountered an error connecting to socket\n");
#ifdef WIN32
								Win32Cleanup();
#endif
								exit(1);
							}								
						}

					}
					
				} else {
					fprintf(stderr,"Protocol not supported\n");
					fprintf(stderr,"Try '%s --help' for more information.\n",progname);
				}
			}
			if (!file_open) {
				break;
			}
		}
	}
#ifdef WIN32
	Win32Cleanup();
#endif
	exit(0);
}


tailor.h
/* tailor.h -- target dependent definitions
 * Copyright (C) 1992-1993 Jean-loup Gailly.
 * This is free software; you can redistribute it and/or modify it under the
 * terms of the GNU General Public License, see the file COPYING.
 */

/* The target dependent definitions should be defined here only.
 * The target dependent functions should be defined in tailor.c.
 */

/* $Id: tailor.h,v 1.1.1.1 2001/11/01 21:13:50 chris Exp $ */

#if defined(__MSDOS__) && !defined(MSDOS)
#  define MSDOS
#endif

#if defined(__OS2__) && !defined(OS2)
#  define OS2
#endif

#if defined(OS2) && defined(MSDOS) /* MS C under OS/2 */
#  undef MSDOS
#endif

#ifdef OS2
#  define PATH_SEP2 '\\'
#  define PATH_SEP3 ':'
#  define MAX_PATH_LEN  260
#  ifdef OS2FAT
#    define NO_MULTIPLE_DOTS
#    define MAX_EXT_CHARS 3
#    define Z_SUFFIX "z"
#    define casemap(c) tolow(c)
#  endif
#  define NO_CHOWN
#  define PROTO
#  define STDC_HEADERS
#  include <io.h>
#  define OS_CODE  0x06
#  define SET_BINARY_MODE(fd) setmode(fd, O_BINARY)
#  ifdef _MSC_VER
#    define HAVE_SYS_UTIME_H
#    define NO_UTIME_H
#    define MAXSEG_64K
#    undef near
#    define near _near
#  endif
#  ifdef __EMX__
#    define HAVE_SYS_UTIME_H
#    define NO_UTIME_H
#    define DIRENT
#    define EXPAND(argc,argv) \
       {_response(&argc, &argv); _wildcard(&argc, &argv);}
#  endif
#  ifdef __BORLANDC__
#    define DIRENT
#  endif
#  ifdef __ZTC__
#    define NO_DIR
#    define NO_UTIME_H
#    include <dos.h>
#    define EXPAND(argc,argv) \
       {response_expand(&argc, &argv);}
#  endif
#endif

#ifdef WIN32 /* Windows NT */
#  define HAVE_SYS_UTIME_H
#  define NO_UTIME_H
#  define PATH_SEP2 '\\'
#  define PATH_SEP3 ':'
#  define SUFFIX_SEP '.'
#endif

#if defined(VAXC) || defined(VMS)
#  define PATH_SEP ']'
#  define PATH_SEP2 ':'
#  define SUFFIX_SEP ';'
#endif

#ifdef AMIGA
#  define PATH_SEP2 ':'
#  define STDC_HEADERS
#endif

#if defined(ATARI) || defined(atarist)
#  ifndef STDC_HEADERS
#    define STDC_HEADERS
#    define HAVE_UNISTD_H
#    define DIRENT
#  endif
#  define ASMV
#  define OS_CODE  0x05
#  ifdef TOSFS
#    define PATH_SEP2 '\\'
#    define PATH_SEP3 ':'
#    define MAX_PATH_LEN  128
#    define NO_MULTIPLE_DOTS
#    define MAX_EXT_CHARS 3
#    define Z_SUFFIX "z"
#    define NO_CHOWN
#    define casemap(c) tolow(c) /* Force file names to lower case */
#    define NO_SYMLINK
#  endif
#endif

#ifdef MACOS
#  define PATH_SEP ':'
#endif

#ifdef __50SERIES /* Prime/PRIMOS */
#  define PATH_SEP '>'
#  define STDC_HEADERS
#endif

#ifndef unix
#  define NO_ST_INO /* don't rely on inode numbers */
#endif


	/* Common defaults */

#ifndef PATH_SEP
#  define PATH_SEP '/'
#endif

#ifndef casemap
#  define casemap(c) (c)
#endif


version.h
/*
 * chop 
 * ( and it's nemesis unchop )
 * Copyright (C) 1999 BingosNET Produktions Ltd
 * Written by Kidney Bingos aka Chris Williams
 * <chris@kidney-bingos.demon.co.uk> 
 * With some components borrowed from the Free Software Foundation
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

#if defined __MINGW32__ 
#define COMPILER "GNU C Compiler mingw32"
#elif defined _MSC_VER
#define COMPILER "Microsoft Visual C++"
#elif defined __BORLANDC__
#define COMPILER "Borland C Compiler"
#elif defined __CYGWIN32__
#define COMPILER "Cygnus GNU C Compiler Win32"
#elif defined __GNUC__ 
#define COMPILER "GNU C Compiler"
#elif defined __HIGHC__
#define COMPILER "NCR High Performance C Compiler"
#else
#define COMPILER "Unknown"
#endif

#if defined CONF_MACHTYPE
#define BUILD_ENV CONF_MACHTYPE
#else
#define BUILD_ENV "Not defined"
#endif


furlutil.h
/*
 * furl 
 * Copyright (C) 1999 BingosNET Produktions Ltd
 * Written by Kidney Bingos aka Chris Williams
 * <chris@kidney-bingos.demon.co.uk> 
 * With some components borrowed from the Free Software Foundation
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

#include <config.h>

#ifdef WIN32
#include <windows.h>
#define ERR_NO WSAGetLastError
#else
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netdb.h>
#include <arpa/inet.h>
#include <unistd.h>
#define SOCKET int
#define INVALID_SOCKET -1
#define SOCKET_ERROR -1
#define ERR_NO errno
#endif

#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#define tolow(c)  (isupper(c) ? (c)-'A'+'a' : (c))    /* force to lower case */

extern char *basename(char *fname);
extern char *strlwr(char *s);
extern void PrintUsage(const char *prog_name);
extern void PrintVersion(const char *prog_name);
extern int parseurl(char *url, char protocol[], char hostname[], char port[], char path[], char method[], char useragent[]);
extern int GetAddress(unsigned long *ipAddr, char hostname[]);
extern SOCKET CreateSocket(unsigned long *ipAddr, char port[]);
extern SOCKET ProcessSocket(SOCKET sock, char path[]);

#ifdef WIN32
extern void Win32Init(void);
extern void Win32Cleanup(void);
#endif


Win32/config.h
/* config.h.  Generated automatically by configure.  */
/* config.h.in.  Generated automatically from configure.in by autoheader.  */

/* Define to empty if the keyword does not work.  */
/* #undef const */

/* Define to `int' if <sys/types.h> doesn't define.  */
/* #undef pid_t */

/* Define if you have the ANSI C header files.  */
#define STDC_HEADERS 1

/* #undef CONF_MACHTYPE */
/* #undef WIN32 */

/* Define if you have the gethostbyname function.  */
/* #undef HAVE_GETHOSTBYNAME */

/* Define if you have the <unistd.h> header file.  */
#define HAVE_UNISTD_H 1

/* Define if you have the resolv library (-lresolv).  */
/* #undef HAVE_LIBRESOLV */

/* Define if you have the socket library (-lsocket).  */
/* #undef HAVE_LIBSOCKET */

/* Name of package */
#define PACKAGE "furl"

/* Version number of package */
#define VERSION "2.1"



furlutil.c
/*
 * furl
 * Copyright (C) 1999 BingosNET Produktions Ltd
 * Written by Kidney Bingos aka Chris Williams
 * <chris@bingosnet.co.uk> 
 * With some components borrowed from the Free Software Foundation
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

#include <ctype.h>
#include <errno.h>
#include <sys/types.h>
#include <stdlib.h>
#include "furlutil.h"
#include "version.h"
#include "tailor.h"

/* ========================================================================
 * Return the base name of a file (remove any directory prefix and
 * any version suffix). For systems with file names that are not
 * case sensitive, force the base name to lower case.
 */
char *basename(fname)
    char *fname;
{
    char *p;

    if ((p = strrchr(fname, PATH_SEP))  != NULL) fname = p+1;
#ifdef PATH_SEP2
    if ((p = strrchr(fname, PATH_SEP2)) != NULL) fname = p+1;
#endif
#ifdef PATH_SEP3
    if ((p = strrchr(fname, PATH_SEP3)) != NULL) fname = p+1;
#endif
#ifdef SUFFIX_SEP
    if ((p = strrchr(fname, SUFFIX_SEP)) != NULL) *p = '\0';
#endif
    if (casemap('A') == 'a') strlwr(fname);
    return fname;
}

/* ========================================================================
 * Put string s in lower case, return s.
 */
char *strlwr(s)
    char *s;
{
    char *t;
    for (t = s; *t; t++) *t = tolow(*t);
    return s;
}

void PrintUsage(const char *prog_name)
{
	printf("Usage: %s [OPTIONS] [URL] [filename]\n",prog_name);
	printf("Where URL is in the format:\n\n");
	printf("      [http://]www.foobar.com[:80][/example.html]\n\n");
	printf("\n");
	printf("-f, --file     specify an input file containing URLs\n");
	printf("-b, --body     output the contents of HTTP request and the header\n");
	printf("-V, --version  output version information and exit\n");
	printf("-h, --help     display this help message and exit\n");
	printf("-i, --ie       pretend to be Internet Explorer\n");
	printf("-m, --mozilla  pretend to be Mozilla\n");
	printf("\nBug reports to furl-bugs@bingosnet.co.uk\n"); 
}

void PrintVersion(const char *prog_name)
{
	printf("%s\n\n", prog_name);
	printf("version:                    %s\n",VERSION);
	printf("compile date:               %s %s\n",__DATE__,__TIME__);
#if defined __GNUC__
	printf("compiler:                   %s %s\n",COMPILER,__VERSION__);
#else
	printf("compiler:                   %s\n",COMPILER);
#endif
	printf("environment:                %s\n",BUILD_ENV);
	printf("author:                     Kidney Bingos (chris@bingosnet.co.uk)\n");
}

int parseurl (char *url, char protocol[], char hostname[], char port[], char path[], char method[], char useragent[])
{
	char *m, *s;

	if (!strncmp(url,"http://",7)) {
		s = url + 7;
		strncpy(protocol,"http://",8);
	} else if (!strncmp(url,"ftp://",6)) {
		return -1;
	} else if (!strncmp(url,"gopher://",9)) {
		return -1;
	} else if (!strncmp(url,"raw://",6)) {
		return -1;
	} else if (!strncmp(url,"mailto://",9)) {
		return -1;
	} else if (!strncmp(url,"telnet://",9)) {
		return -1;
	} else {
		s = url;
		strncpy(protocol,"http://",8);
	}

	for (m=hostname;s[0]!=':'&&s[0]!='/'&&s[0];) {
		*m++ = *s++;
	}
	m[0] = '\0';
	m = port;
	if (s[0] == ':') {
		s++;
		do {
			*m++ = *s++;
		} while (isdigit(s[0]));
		*m = 0;
	} else {
		strcpy(m,"80");
	}
	sprintf(path,"%s %s HTTP/1.1\r\nConnection: close\r\nHost: %s\r\nAccept: */*\r\nUser-Agent: %s\r\n\r\n", method, (*s) ? s : "/", hostname, useragent);
	return 0;
}


int GetAddress (unsigned long *ipAddr, char hostname[])
{
	struct hostent *ent;
	unsigned long a,b,c,d;

	if (sscanf(hostname,"%ld.%ld.%ld.%ld", &a, &b, &c, &d) == 4) {

		*ipAddr = (a << 24) | (b << 16) | (c << 8) | d;
	} else {
		ent = gethostbyname(hostname);
		if (!ent) {
			return -1;
		}
		*ipAddr = ntohl(**((unsigned long**)ent->h_addr_list));
	}
	return 0;
}

SOCKET CreateSocket (unsigned long *ipAddr, char port[])
{

	struct sockaddr_in sa;
	SOCKET sock;

	sock = socket (AF_INET,SOCK_STREAM,IPPROTO_TCP);
	if (sock == INVALID_SOCKET) {
		return INVALID_SOCKET;
	}

	sa.sin_family = AF_INET;
	sa.sin_addr.s_addr = htonl(*ipAddr);
	sa.sin_port = htons(atoi(port));

	if (connect(sock,(struct sockaddr*)&sa,sizeof(sa)) == SOCKET_ERROR) {
#ifdef WIN32
		closesocket(sock);
#else
		close(sock);
#endif
		return INVALID_SOCKET;
	}

	return sock;

}

SOCKET ProcessSocket (SOCKET sock, char path[])
{

	char buffer[64*1024];
	SOCKET n;

	if (send(sock,path,strlen(path),0) == SOCKET_ERROR) {
		perror("Ooops:");
		return SOCKET_ERROR;
	} else {
		
		while ((n = recv(sock,buffer,sizeof(buffer),0)) > 0 ) {
			buffer[n] = 0;
			if (fputs(buffer,stdout) == EOF ) {
				fprintf(stderr,"Bad Magic. Couldn\'t write to STDOUT\n");
#ifdef WIN32
				closesocket(sock);
#else
				close(sock);
#endif
				return SOCKET_ERROR;
			}
		}
		if (n < 0) {
			return SOCKET_ERROR;
		} else {
			return 0;
		}
	}
}

#ifdef WIN32

void Win32Init (void)
{
	WSADATA wsaData;

	if (WSAStartup(MAKEWORD(2,0), &wsaData)) {
		fprintf(stderr,"Error initialising winsock.dll\n");
		exit(1);
	}
}

void Win32Cleanup (void)
{
	WSACleanup();
}

#endif
